# <p style="background-color:#718355;font-family:newtimeroman;color:#FFFFFF;font-size:150%;text-align:center;border-radius:10px 10px;">Análisis de la Personalidad del Cliente</p>
---

[![proyectofinalpeque-o.png](https://i.postimg.cc/Y0HvGmRd/proyectofinalpeque-o.png)](https://postimg.cc/JDprS09X)

# <p style="background-color:#718355;font-family:newtimeroman;color:#FFFFFF;font-size:150%;text-align:center;border-radius:10px 10px;">Tabla de Contenidos</p>

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#sec_1">Planteamiento del problema</a>  
2. <a href="#sec_2">Exploración del set de datos</a>  
3. <a href="#sec_3">Limpieza del set de datos</a>  
4. <a href="#sec_4">Análisis de los datos</a>  
5. <a href="#sec_5">Preprocesamiento</a>  
6. <a href="#sec_6">Clasificación de Clientes</a> 
7. <a href="#sec_7">Reporte final</a>   

</font>
</div>

<a id='sec_1'></a>
## `1. Planteamiento del problema`


El análisis de la personalidad del cliente ayuda a una empresa a modificar su producto en función de sus clientes objetivo de diferentes tipos de segmentos de clientes. Por ejemplo, en lugar de gastar dinero para comercializar un nuevo producto para cada cliente de la base de datos de la empresa, una empresa puede analizar qué segmento de clientes tiene más probabilidades de comprar el producto y luego comercializarlo sólo en ese segmento en particular.

**1.2. Recursos disponibles**
- [Set de datos Marketing_Camapaña](marketing_campaign.csv)

**1.3. Diccionario de datos disponibles**
| **Clave** | **Descripción** | **Tipo de dato** |
|---|---|---|
| ID | identificador único del cliente | Int |
| Year_Birth | Año de nacimiento | Int |
| Education | Nivel de educación | Object |
| Marital_Status | Estado civil | Object |
| Income | Ingreso anual | Float |
| Kidhome | Niños en el hogar | Int |
| Teenhome | Adolescentes en el hogar | Int |
| Dt_Customer | Fecha de alta del cliente | Object |
| Recency | Número de días desde la última compra| Int |
| Complain | 1 si el cliente se quejó en los últimos 2 años, 0 en caso contrario | Int |
| MntWines | Monto gastado en vino en los últimos 2 años | Int |
| MntFruits | Monto gastado en frutas en los últimos 2 años | Int |
| MntMeatProducts | Monto gastado en carne en los últimos 2 años | Int |
| MntFishProducts | Monto gastado en pescado en los últimos 2 años | Int |
| MntSweetProducts | Monto gastadoa en dulces en los últimos 2 años | Int |
| MntGoldProds | Monto gastado en oro en los últimos 2 años | Int |
| NumDealsPurchases | Compras realizadas con descuento | Int |
| AcceptedCmp1 | 1 aceptó la oferta en la primera campaña, 0 en caso contrario | Int |
| AcceptedCmp2 | 1 aceptó la oferta en la segunda campaña, 0 en caso contrario | Int |
| AcceptedCmp3 | 1 aceptó la oferta en la tercera campaña, 0 en caso contrario | Int |
| AcceptedCmp4 | 1 aceptó la oferta en la cuarta campaña, 0 en caso contrario | Int |
| AcceptedCmp5 | 1 aceptó la oferta en la quinta campaña, 0 en caso contrario | Int |
| Response | 1 si el cliente aceptó la oferta en la última campaña, 0 en caso contrario | Int |
| NumWebPurchases | Compras realizadas a través del sitio web de la empresa | Int |
| NumCatalogPurchases | Compras realizadas mediante un catálogo | Int |
| NumStorePurchases | Compras realizadas directamente en tiendas | Int |
| NumWebVisitsMonth | Visitas al sitio web de la empresa en el último mes | Int |

<a id='sec_2'></a>
## `2. Exploración del set de datos`

El objetivo de esta fase es poder familiarizarnos con la información contenida en el set de datos.

1. Leer el set de datos
2. Entender tamaño, columnas y filas del set de datos
3. Entender qué tipos de datos tenemos en cada columna
4. Verificar si hay datos faltantes y valores duplicados

**2.1. Leer el set de datos**

In [12]:
import pandas as pd

In [14]:
df = pd.read_csv(r"marketing_campaign.csv",sep='\t')
df.head(5)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  04-09-2012       58       635  ...                  7             0   
1  08-03-2014       38        11  ...                  5             0   
2  21-08-2013       26       426  ...                  4             0   
3  10-02-2014       26        11  ...                  6             0   
4  19-01-2014       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

**2.2. Entender tamaño, columnas y filas del set de datos**

In [15]:
print(f'El número de filas es: {df.shape[0]}')
print(f'El número de columnas es: {df.shape[1]}')

El número de filas es: 2240
El número de columnas es: 29


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

**2.3. Entender qué tipos de datos tenemos en cada columna**

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

>**Observación**
> - Al poder observar a mas detalle el tipo de dato que tiene cada columna, se pudo ver que exiten datos de tipo `int` , `float` y `Object`.
> - Se puede observar que tenemos la `fecha de alta del cliente` como tipo de dato `Object`, el cual en caso se requiera analizarlo se podria cambiar a tipo de dato`Date`

**2.4. Verificar si hay datos faltantes y valores duplicados**

In [18]:
df.duplicated().sum()

0

>**Observación**
> - Al observar se pudo evidenciar que no hay valores duplicados.

In [19]:
df.isna().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

>**Observación**
> - Al observar se pudo evidenciar que  hay valores faltantes en la columna `Income`.

<a id='sec_3'></a>
## `3. Limpieza de los datos`

Etapas de limpieza:

1. Revisión y eliminación datos duplicados
2. Eliminación de columnas innecesarias
3. Revisión y imputación de valores faltantes
4. Creación de nuevas columnas(caracteristicas)
5. Revisión de columnas categoricas
6. Revisión o eliminación de valores outliers

Implementemos cada fase teniendo en cuenta las observaciones que hemos hecho hasta el momento.

**3.1 Revisión y eliminación datos duplicados**

In [20]:
df = df.drop_duplicates()

In [21]:
df.duplicated().sum()

0

>**Observación**
> - Al observar se pudo evidenciar que ya no se cuenta con valores duplicados

**3.2. Eliminación de columnas innecesarias**

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

>**Observación**
> - De las observaciones hechas anteriormente concluimos que se pueden eliminar la columna: `ID`
> - Esto lo haremos con el método `drop`

In [24]:
df = df.drop(columns=['ID'])
df

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0           1957  Graduation         Single  58138.0        0         0   
1           1954  Graduation         Single  46344.0        1         1   
2           1965  Graduation       Together  71613.0        0         0   
3           1984  Graduation       Together  26646.0        1         0   
4           1981         PhD        Married  58293.0        1         0   
...          ...         ...            ...      ...      ...       ...   
2235        1967  Graduation        Married  61223.0        0         1   
2236        1946         PhD       Together  64014.0        2         1   
2237        1981  Graduation       Divorced  56981.0        0         0   
2238        1956      Master       Together  69245.0        0         1   
2239        1954         PhD        Married  52869.0        1         1   

     Dt_Customer  Recency  MntWines  MntFruits  ...  NumWebVisitsMonth  \
0     04-09-2012       58       635         88  ...                  7   
1     08-03-2014       38        11          1  ...                  5   
2     21-08-2013       26       426         49  ...                  4   
3     10-02-2014       26        11          4  ...                  6   
4     19-01-2014       94       173         43  ...                  5   
...          ...      ...       ...        ...  ...                ...   
2235  13-06-2013       46       709         43  ...                  5   
2236  10-06-2014       56       406          0  ...                  7   
2237  25-01-2014       91       908         48  ...                  6   
2238  24-01-2014        8       428         30  ...                  3   
2239  15-10-2012       40        84          3  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 28 columns]

>**Observación**
> - Como se puede observar se elimino la columna `ID`, debido a que no es necesario para nuestro modelo por queremos hacer un modelo de clustering y no seria lo recomendable tener un ID, perjudica a nuestro modelo

**3. Revisión y imputación de valores faltantes**

In [31]:
df.isna().sum()

Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
Age                     0
Children                0
dtype: int64

In [32]:
df_faltantes = df.isnull().sum()
porcentaje_faltantes = (df_faltantes / df.shape[0]) * 100
porcentaje_faltantes

Year_Birth             0.000000
Education              0.000000
Marital_Status         0.000000
Income                 1.071429
Kidhome                0.000000
Teenhome               0.000000
Dt_Customer            0.000000
Recency                0.000000
MntWines               0.000000
MntFruits              0.000000
MntMeatProducts        0.000000
MntFishProducts        0.000000
MntSweetProducts       0.000000
MntGoldProds           0.000000
NumDealsPurchases      0.000000
NumWebPurchases        0.000000
NumCatalogPurchases    0.000000
NumStorePurchases      0.000000
NumWebVisitsMonth      0.000000
AcceptedCmp3           0.000000
AcceptedCmp4           0.000000
AcceptedCmp5           0.000000
AcceptedCmp1           0.000000
AcceptedCmp2           0.000000
Complain               0.000000
Z_CostContact          0.000000
Z_Revenue              0.000000
Response               0.000000
Age                    0.000000
Children               0.000000
dtype: float64

>**Observación**
> - Como se puede observar se tiene 24 valores faltantes en la columna `Income` es decir el 1.07 % de toda nuestra data

In [34]:
df[['Income']].describe().T

count          mean           std     min      25%      50%      75%  \
Income  2216.0  52247.251354  25173.076661  1730.0  35303.0  51381.5  68522.0   

             max  
Income  666666.0

**3.4. Creación de nuevas columnas(caracteristicas)**

- *Creación de una columna Edad*

In [26]:
df['Age'] = 2024 - df['Year_Birth']

>**Observación**
> - Se creo esta nueva columna Age con el fin de poder hacer una analisis más sencillo en rango de las edades de los clientes

- *Creación de una columna niñ@s*

In [27]:
df["Children"]= df["Kidhome"]+ df["Teenhome"]

>**Observación**
> - Se creo esta nueva columna Children con el fin de poder hacer una analisis más sencillo del total de niños y adolescente que vive el cliente

**3.5. Revisión de columnas categoricas**

In [36]:
for column in df.select_dtypes(include = 'object'):
    print('-'*30)
    print(f'** {column} ***')
    print(df[column].value_counts(normalize=True).round(2))
    print('-'*30)

------------------------------
** Education ***
Education
Graduation    0.50
PhD           0.22
Master        0.17
2n Cycle      0.09
Basic         0.02
Name: proportion, dtype: float64
------------------------------
------------------------------
** Marital_Status ***
Marital_Status
Married     0.39
Together    0.26
Single      0.21
Divorced    0.10
Widow       0.03
Alone       0.00
Absurd      0.00
YOLO        0.00
Name: proportion, dtype: float64
------------------------------
------------------------------
** Dt_Customer ***
Dt_Customer
31-08-2012    0.01
12-09-2012    0.00
14-02-2013    0.00
12-05-2014    0.00
20-08-2013    0.00
              ... 
05-08-2012    0.00
18-11-2012    0.00
25-05-2013    0.00
14-04-2013    0.00
09-01-2014    0.00
Name: proportion, Length: 663, dtype: float64
------------------------------


**3.6. Revisión o eliminación de valores outliers**